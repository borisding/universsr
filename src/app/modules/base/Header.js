import React from 'react';
import { NavLink } from 'react-router-dom';
import { routes } from '@app/routes';
import { active } from './styles/Header.scss';

const Header = () => (
  <header>
    <nav styleName="menu">
      <ul>
        {routes.map(
          (route, index) =>
            route.menu && (
              <li key={index}>
                <NavLink
                  to={route.path}
                  exact={route.exact === true}
                  activeClassName={active}
                >
                  {route.menu}
                </NavLink>
              </li>
            )
        )}
      </ul>
    </nav>
  </header>
);

export default Header;
