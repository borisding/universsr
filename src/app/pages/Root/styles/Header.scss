@import '@common/styles/variables';

header {
  background-color: lighten($color-black, 10%);
  color: $color-white;
  display: block;
  height: $header-height;
  left: 0;
  margin: 0;
  padding: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 100;
}

.menu {
  .active,
  %active {
    background-color: lighten($color-black, 25%);
    text-decoration: underline;
  }

  ul {
    height: $header-height;
    list-style: none;
  }

  li {
    display: inline-block;
    padding: 0;
    text-align: center;
  }

  li > a {
    color: $color-white;
    display: inline-block;
    height: $header-height;
    line-height: $header-height;
    min-width: 60px;
    padding: 0 $base-padding;
    text-decoration: none;
  }

  li > a:hover {
    @extend %active;
  }
}
