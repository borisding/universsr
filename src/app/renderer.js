import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
