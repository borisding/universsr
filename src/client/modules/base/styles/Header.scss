@import '~@styles/variables';

$height: 60px;
$width: 60px;

header {
  background-color: lighten($color-black, 10%);
  color: $color-white;
  height: $height;
  left: 0;
  margin: 0;
  padding: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 100;
}

.menu {
  .active,
  %active {
    background-color: lighten($color-black, 25%);
    text-decoration: underline;
  }

  ul {
    height: $height;
    list-style: none;
  }

  li {
    display: inline-block;
    height: $height;
    padding: 0;
    text-align: center;
  }

  li > a {
    color: $color-white;
    display: inline-block;
    height: $height;
    line-height: $height;
    min-width: $width;
    padding: 0 $base-padding;
    text-decoration: none;
  }

  li > a:hover {
    @extend %active;
  }
}
