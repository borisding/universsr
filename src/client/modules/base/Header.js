import React from 'react';
import { NavLink } from 'react-router-dom';
import { routes } from '@client/routes';
import styles from './styles/Header.scss';

const Header = () => (
  <header>
    <nav className={styles.menu}>
      <ul>
        {routes.map(
          (route, index) =>
            route.menu && (
              <li key={index}>
                <NavLink
                  to={route.path}
                  exact={route.exact === true}
                  activeClassName={styles.active}
                >
                  {route.menu}
                </NavLink>
              </li>
            )
        )}
      </ul>
    </nav>
  </header>
);

export default Header;
