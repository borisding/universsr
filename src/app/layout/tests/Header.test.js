import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Header from '../Header';

describe('Header', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
