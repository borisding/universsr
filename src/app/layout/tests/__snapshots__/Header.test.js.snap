// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header>
  <nav
    className="menu"
  >
    <ul>
      <li>
        <a
          aria-current="page"
          className="active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Home
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/todos"
          onClick={[Function]}
        >
          Todos
        </a>
      </li>
    </ul>
  </nav>
</header>
`;
